docs/security/index.md

```diff
@@ -0,0 +1,400 @@
+`/opt/telicent/security/plugins/` - that all applications **MUST** add to their classpath if it exists.  Plugins
+**SHOULD** be mounted as read-only files to minimize the possibility of any tampering.
+
+The proposed plugin mounting mechanism is that us, or the customer, builds a container image that contains their plugin
+(and its dependencies) and uses a K8S init container to copy these into an `emptyDir` volume that will be shared with
+the main application container.  This init container can be injected into all relevant application manifests by way of
+common `kustomize` patches.
+
+Telicent will provide a default patch as part of our base manifests that injects our [Default Plugin](#default-plugin).
+
+## Interfaces
+
+### `SecurityLabels<?>`
+
+`SecurityLabels` is a generic type that holds security labels, it holds the `encoded()` byte sequence as a `byte[]`, a
+`schema()` identifier as a `short`, and its `decodedLabels()` method provides access to the concrete labels type a
+security plugin uses for its implementation.  It also has a `toDebugString()` method that allows implementations to
+provide a human-readable representation of the labels for debugging purposes, note that this differs from the standard
+`toString()` method in that it may want to translate from whatever encoding schema has been used to something a
+developer/system operator can understand, whereas `toString()` should just provide a quick view of the labels e.g.
+schema and encoded data length.
+
+The `AbstractSecurityPrimitive` class provides a basic implementation of most of the interface allowing implementations
+to focus on their implementation detail.
+
+### `Entitlements<?>`
+
+`Entitlements` is a generic type that holds user entitlements, it holds the `encoded()` byte sequence as a `byte[]`, a
+`schema()` identifier as a `short`, and its `decodedEntitlements()` method provides access to the concrete entitlements
+type a security plugin uses for its implementation.  Again the `AbstractSecurityPrimitive` class provides a basic
+implementation of most of the interface.
+
+### `SecurityPlugin`
+
+`SecurityPlugin` is the top level entrypoint API by which applications interact with the Security Plugin API.  Primarily
+this is just a place for applications to obtain instances of the other interfaces they need to carry out data processing
+and access control tasks.
+
+An implementation should generally make the instances of the other interfaces singletons where the API allows it e.g.
+there should be a single shared `SecurityLabelsParser`.  However implementations **MUST** be aware that they will be
+used in multi-threaded environments thus any singleton instance they return **MUST** be thread-safe.
+
+Note that some of the interfaces a plugin provides instances of have specific lifecycles associated with them, e.g.
+`Authorizer`, so for those implementations **MUST** ensure they respect the defined lifecycle of that interface.
```

Comment:
- [ ] TODO: Consider making interfaces with short lifecycles clearer by making them `AutoCloseable` so the JVM will complain if applications don't consume and `close()` these appropriately.